cluster-autoscaler/cloudprovider/test/test_cloud_provider.go

```diff
@@ -330,7 +333,7 @@ func (tng *TestNodeGroup) TemplateNodeInfo() (*schedulercache.NodeInfo, error) {
 	}
 	template, found := tng.cloudProvider.machineTemplates[tng.id]
 	if !found {
-		return nil, fmt.Errorf("No template declared for %s", tng.id)
+		return nil, cloudprovider.ErrNotImplemented
```

Comment:
I made this change because I want to be able to use NodeInfo without TemplateNodeInfo and NodeInfo with TemplateNodeInfo in one unit test. In previous version if I define machine template for one NodeInfo then I need to define it for all NodeInfos. Is returning an error so crucial? This is an util class for testing.

Reply:
This error is missleading. The implementation for this method is provided. The only thing is that the author of the test didn't provide the template for one of the node groups.

Reply:
The edge case in test appears to be a nodegroup with only unready nodes. In this case, GCE cloud provider attempts to retrieve instance template from MIG, and any error generated at this point comes directly from cloud provider. Not adding template to test cloud provider simulates such error. I think changing this would reduce the mocks usefulness by taking away the ability to test this scenario.

But I agree that failing at this point is not necessarily the desired behavior, as a single unhealthy node group could block scale up. I think fixing it would require change to GetNodeInfosForGroups, as indicated here: 
https://github.com/kubernetes/autoscaler/blob/f9cabf3a1a648af2caaaec87dea781446f15f13c/cluster-autoscaler/core/utils.go#L158